### PR TITLE
Module 1

### PR DESCRIPTION
**Pull request by dan1dr**
Adjunto el assignment del Módulo 1

**Partes:**

- Import librerias y definición de variables
- call_api(): Llamar API
- get_data_meteo_api(): Extraer datos
- clean_data() (no usada): Limpiar datos de un dict una vez extraídos (pensaba manejar diccionarios inicialmente antes de ver una solución con pandas)
- process_data(): Procesar los datos
- plot_data(): Representación de los datos extraída del código ejemplo

**Dudas y comentarios:**

1. He creado la branch con el problema ya iniciado en vez de crearlo desde cero (mea culpa).
2. La parte de testing es donde me cuesta más interiorizar qué testear y cómo. Me faltaría aún digerir la parte de mock testing añadida por Guille.
3. Cuál sería el modo de afrontar este assignment y hacer el prototype (i.e. _opción 1)_ me pongo en VS Code a desarrollarlo del tirón y ejecuto con CTRL + F5 alineando las funciones en main o quizás mejor usar shell y correr "python x.py"? _opción 2)_ voy corriendo funcion a funcion para ver que hace lo que quiero con variables locales y así hasta el final? _opción_3)_ lo hago desde la extensión de notebooks incluida en VS Code y después me muevo a un script.py? Feel free de dar unas recomendaciones a la hora de trabajar con VS Code y desarrollo de código. Concretamente, cuál es la secuencia lógica en la que debería afrontar un assignment y pasos a seguir "idealmente" a pesar de que puedan ser criterios muy personales. Estoy acostumbrado a experimentar y explorar con notebooks y está claro que no te da las mejores prácticas a la hora de pasarte a desarrollar todo en un IDE.

*He tenido bastantes problemas a la hora de la instalación de paquetes con SSL y certificados al trabajar desde ordenador corporativo. Deshabilitando SSL, actualizando certs, cambiando git config de http.backend, VPNs, etc... Igualmente para usar la API (por defecto, verify = True deniega acceso)

climate_evolution

### CODE REVIEW

File: src/module_1/module_1_meteo_api.py [hunk 1]
```diff
@@ -3,34 +3,45 @@
...
+import urllib3
+# import json
+# import numpy as np
+
+
+# Disable the InsecureRequestWarning due to using verify=False from corporate laptop
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
 
 """Define global variables"""
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: Use # for one line comments

File: src/module_1/module_1_meteo_api.py [hunk 2]
```diff
@@ -3,34 +3,45 @@
...
+
 API_URL = "https://climate-api.open-meteo.com/v1/climate?"
 
 COOL_OFF_TIME = 5
 MAX_RETRY_ATTEMPTS = 3
 
 COORDINATES = {
-"Madrid": {"latitude": 40.416775, "longitude": -3.703790},
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py:18]: Use a Python formatter in VS code which should handle this automatically

File: src/module_1/module_1_meteo_api.py [hunk 3]
```diff
@@ -3,34 +3,45 @@
...
-"London": {"latitude": 51.507351, "longitude": -0.127758},
-"Rio": {"latitude": -22.906847, "longitude": -43.172896},
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
 }
 
 VARIABLES = "temperature_2m_mean,precipitation_sum,soil_moisture_0_to_10cm_mean"
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: Even if you are gonna use this as an string later, use a list here. A raw string is error prone, hard to read and hard to modify

File: src/module_1/module_1_meteo_api.py [hunk 4]
```diff
@@ -3,34 +3,45 @@
...
 
-#All models have data for the 3 variables except soil moisture, which is only provided by MRI_AGCM3 and EC_Earth
-MODELS = "CMCC_CM2_VHR4,FGOALS_f3_H,HiRAM_SIT_HR,MRI_AGCM3_2_S,EC_Earth3P_HR,MPI_ESM1_2_XR,NICAM16_8S"
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py:26]: Same as before

File: src/module_1/module_1_meteo_api.py [hunk 5]
```diff
@@ -88,81 +111,96 @@ def clean_data(raw_data):
...
+        # Save each variable with its corresponding mean and std (axis=1 per each row)
+        # We have decided to compute the mean of all models for one day
+        # e.g. mean of all precipitation_sum for Madrid on 1950-01-01
+        # We could have done the mean of all occurences during one year for one model
+        calculated_df[f"{var}_mean"] = data[idxs].mean(axis=1)
+        calculated_df[f"{var}_std"] = data[idxs].std(axis=1)
+    return calculated_df
+
+
+def plot_data(data):
```
Comment (guillerbf) [inline src/module_1/module_1_meteo_api.py:131]: Si cogéis el código que subimos los profesores y lo poneis en vuestra PR no tiene mucha gracia. Además, este no es necesariamente el mejor código ni la mejor solución. La gracia del code review es que sea el vuestro para que os lo podamos revisar y aprendáis en el proceso.

File: src/module_1/module_1_meteo_api.py [hunk 6]
```diff
@@ -3,34 +3,45 @@
...
 VARIABLES = "temperature_2m_mean,precipitation_sum,soil_moisture_0_to_10cm_mean"
 
-#All models have data for the 3 variables except soil moisture, which is only provided by MRI_AGCM3 and EC_Earth
-MODELS = "CMCC_CM2_VHR4,FGOALS_f3_H,HiRAM_SIT_HR,MRI_AGCM3_2_S,EC_Earth3P_HR,MPI_ESM1_2_XR,NICAM16_8S"
+# All models have data for the 3 variables except soil moisture
+# which is only provided by MRI_AGCM3 and EC_Earth
+MODELS = "CMCC_CM2_VHR4,FGOALS_f3_H,HiRAM_SIT_HR,MRI_AGCM3_2_S,EC_Earth3P_HR,MPI_ESM1_2_XR,NICAM16_8S" # noqa
+
 
 """Define auxiliar functions"""
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: Ensure consistency between """  (docstring) and # (comments) when commenting: https://stackoverflow.com/questions/19074745/docstrings-vs-comments. Ideally, if you comment single lines, use comments, if you comment function headers use docstring

File: src/module_1/module_1_meteo_api.py [hunk 7]
```diff
@@ -3,34 +3,45 @@
...
+
 def call_api(url):
-#Need to add the verify=False as working from my corporate laptop. Tried to authenticate SSL by changing
-#lots of different things but issue still arises. Disabled SSL but only works from the office - not the best practice I know
+    # Need to add the verify=False as working from my corporate laptop.
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: Use docstring here

File: src/module_1/module_1_meteo_api.py [hunk 8]
```diff
@@ -3,34 +3,45 @@
...
+    # Tried to authenticate SSL by changing lots of config, disabled SSL, etc
+    # It only works from the office - not the best practice I know # noqa
     try:
-        #to-do: add the cool off
+        # to-do: add the cool off
         response = requests.get(url, verify=False)
         if response.status_code == 200:
             print("\nConnected Succesfully!")
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: Any reason to use new line at the beginning of the print / log? I think at the end makes more sense

Comment (gsantam) [reply src/module_1/module_1_meteo_api.py]: Don't mix prints and loggins. You can use `logging.info` instead of "prints" and redirect the output to the console

File: src/module_1/module_1_meteo_api.py [hunk 9]
```diff
@@ -39,25 +50,28 @@ def call_api(url):
             logging.error(f"\nAPI request failed with Code: {response.status_code}")
             return None
     except requests.exceptions.RequestException as exception:
-            logging.error(f"\nAPI request failed with Exception: {exception}")
-            return None
+        logging.error(f"\nAPI request failed with Exception: {exception}")
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: Output the requested url

File: src/module_1/module_1_meteo_api.py [hunk 10]
```diff
@@ -39,25 +50,28 @@ def call_api(url):
...
+        return None
 
 
 def get_data_meteo_api(city, from_year, until_year):
     coordinates = COORDINATES.get(city)
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py:88]: Nitpick: I prefere to do `if city not in COORDINATES`

File: src/module_1/module_1_meteo_api.py [hunk 11]
```diff
@@ -39,25 +50,28 @@ def call_api(url):
...
+    # Create the final URL
+    url = (
+        f"{API_URL}latitude={lat}&longitude={long}&start_date={from_year}"
+        f"-01-01&end_date={until_year}-12-31&models={MODELS}&daily={VARIABLES}"
+    )
+
+    # Define a num of max attempts for calling again
+    retry_count = 0
 
     while retry_count < MAX_RETRY_ATTEMPTS:
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: don't handle this here, this logic should be in `call_api`. Same for everything which has to do with `COOL_OFF_TIME`

File: src/module_1/module_1_meteo_api.py [hunk 12]
```diff
@@ -70,16 +84,25 @@ def get_data_meteo_api(city, from_year, until_year):
     print("You reached the number of maximum attempts. Stopping the execution")
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: Use logging with `WARNING`

File: src/module_1/module_1_meteo_api.py [hunk 13]
```diff
@@ -70,16 +84,25 @@ def get_data_meteo_api(city, from_year, until_year):
...
     sys.exit()
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: don't use this, use return

File: src/module_1/module_1_meteo_api.py [hunk 14]
```diff
@@ -70,16 +84,25 @@ def get_data_meteo_api(city, from_year, until_year):
...
 
+
 def clean_data(raw_data):
+    """
+    (Not used)
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: why "not used"? lol

File: src/module_1/module_1_meteo_api.py [hunk 15]
```diff
@@ -70,16 +84,25 @@ def get_data_meteo_api(city, from_year, until_year):
...
+    Function defined for cleaning None values inside a dictionary
+    """
     cleaned_data = {}
     for key, value in raw_data.items():
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: This assumes that `raw_data` is a `dict`, either verify it or use a type hint in the signature of the function

File: src/module_1/module_1_meteo_api.py [hunk 16]
```diff
@@ -70,16 +84,25 @@ def get_data_meteo_api(city, from_year, until_year):
...
         if value is not None:
             if isinstance(value, dict):
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: I think `type(value) is dict` is better here. This logic is not exactly the same as your `isinstance(value, dict)` since the later would be True also in the case that the class of `value` extends the class `dict`, but this doesn't make much sense in this context

File: src/module_1/module_1_meteo_api.py [hunk 17]
```diff
@@ -88,81 +111,96 @@ def clean_data(raw_data):
                 cleaned_data[key] = value
     return cleaned_data
 
+
 def process_data(data):
+    """
+    Reads a df and for each city calculates mean + std for each variable
+    """
+    calculated_df = data[["city", "time"]].copy()
+    for var in VARIABLES.split(","):
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: You don't need the split if you follow my above suggestion

File: src/module_1/module_1_meteo_api.py [hunk 18]
```diff
@@ -70,16 +84,25 @@ def get_data_meteo_api(city, from_year, until_year):
...
 
+
 def clean_data(raw_data):
+    """
+    (Not used)
+    Function defined for cleaning None values inside a dictionary
+    """
     cleaned_data = {}
     for key, value in raw_data.items():
         if value is not None:
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py]: You are checking the same condition independently for raw numbers, lists and dicts. Could you check it once jusr at the end? This may force you to change your entire logic a little bit but it's fine

File: tests/module_1/test_meteo_api.py [hunk 19]
```diff
@@ -1,7 +1,42 @@
...
+import pandas as pd
 
-from src.module_1.module_1_meteo_api import main
+from src.module_1.module_1_meteo_api import process_data, VARIABLES
 
 
-def test_main():
-    raise NotImplementedError
\ No newline at end of file
+def test_process_data():
```
Comment (gsantam) [inline tests/module_1/test_meteo_api.py:10]: Add tests for other functions, specially for `clean_data`

**Review by gsantam — CHANGES_REQUESTED** [review]
Por favor, para la siguiente asegúrate de que no has hecho push a `main` sin crear una PR antes.
Falta algún test. Para interiorizar esto, piensa en los edge cases que pueden encontrar tus funciones. La función `clean_data` es un buen ejemplo para empezar, asegúrate que testeas los casos en los que tienes diccionarios nesteados, listas dentro de diccionarios, listas dentro de listas...
Sobre el desarrollo, a mí siempre me ha gustado combinar Jupyter con VS code. En VS code implemento el código limpio que debería ir en la PR, en Jupyer voy haciendo pruebas rápidas. Por ejemplo, a la hora de deserializar el JSON y hacer la limpieza es muy buena idea probar con Jupyter primero, ya que te vas a encontrar un montón de casos que tienes que ir corrigiendo "visualmente" y una vez que tengas la función definitiva la pones en el VS code e implementas de paso los tests con los casos que has ido probando en Jupyter.

File: src/module_1/module_1_meteo_api.py (hunk 2 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:18]: Noted down (run `make lint` from cmd for theses cases)

File: src/module_1/module_1_meteo_api.py (hunk 1 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:17]: Solved

File: src/module_1/module_1_meteo_api.py (hunk 4 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:26]: Solved

File: src/module_1/module_1_meteo_api.py (hunk 6 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:37]: Solved

File: src/module_1/module_1_meteo_api.py (hunk 8 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:47]: 1. Initially used \n at the beginning to separate logs from the first warning triggered by SSL verify=false. Once it has been disabled, you're right - no need to use it here.
2. Solved. I've configured the logging system just before the function for that (now I receive the logs in the shell :D). It is set to `level=logging.INFO`. I understand that "DEBUG" is for max. verbosity. What is the common value used there when deployed to PROD?

File: src/module_1/module_1_meteo_api.py (hunk 9 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:53]: Done

File: src/module_1/module_1_meteo_api.py (hunk 10 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:58]: Solved. Please go ahead with future nitpicks 🤓

File: src/module_1/module_1_meteo_api.py (hunk 11 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: Solved both

File: src/module_1/module_1_meteo_api.py (hunk 14 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:90]: I started manipulating dicts but later on found out that it wasn't a good idea, so I decided to leave it there in case it were useful. So I will delete the full parts not finally used

File: src/module_1/module_1_meteo_api.py (hunk 11 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: Additional question: If I run the code I obtain 3 times the print "INFO - API request connected successfully" as we're iterating through 
```
for city, coord in COORDINATES.items():
        data.append(
            pd.DataFrame(get_data_meteo_api(city, 1950, 2050)["daily"]).assign(
                city=city
            )
```
Is this the ideal expected behaviour (i.e. one success log per city request) or would be better idea just one-time log?

File: src/module_1/module_1_meteo_api.py (hunk 17 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:120]: Solved

File: src/module_1/module_1_meteo_api.py (hunk 3 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: src/module_1/module_1_meteo_api.py (hunk 13 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: tests/module_1/test_meteo_api.py (hunk 19 above)
Comment (dan1dr) [reply tests/module_1/test_meteo_api.py:10]: Since I deleted 'clean_data' for not being used, I think it would be a good idea to mock responses using 'call_api'/'get_data_meteo_api' funcs. Rest of comments are already fixed @gsantam

File: src/module_1/module_1_meteo_api.py (hunk 1 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: Solved

File: src/module_1/module_1_meteo_api.py (hunk 2 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:18]: Noted down (run `make lint` from cmd for theses cases)

File: src/module_1/module_1_meteo_api.py (hunk 3 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: src/module_1/module_1_meteo_api.py (hunk 4 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:26]: Solved

File: src/module_1/module_1_meteo_api.py (hunk 6 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: src/module_1/module_1_meteo_api.py (hunk 7 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: src/module_1/module_1_meteo_api.py (hunk 8 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: 1. Initially used \n at the beginning to separate logs from the first warning triggered by SSL verify=false. Once it has been disabled, you're right - no need to use it here.
2. Solved. I've configured the logging system just before the function for that (now I receive the logs in the shell :D). It is set to `level=logging.INFO`. I understand that "DEBUG" is for max. verbosity. What is the common value used there when deployed to PROD?

File: src/module_1/module_1_meteo_api.py (hunk 9 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: src/module_1/module_1_meteo_api.py (hunk 10 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py:88]: Solved. Please go ahead with future nitpicks 🤓

File: src/module_1/module_1_meteo_api.py (hunk 12 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: src/module_1/module_1_meteo_api.py (hunk 13 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: src/module_1/module_1_meteo_api.py (hunk 14 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: I started manipulating dicts but later on found out that it wasn't a good idea, so I decided to leave it there in case it were useful. So I will delete the full parts not finally used

File: src/module_1/module_1_meteo_api.py (hunk 17 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: solved

File: src/module_1/module_1_meteo_api.py (hunk 11 above)
Comment (dan1dr) [reply src/module_1/module_1_meteo_api.py]: .

File: tests/module_1/test_meteo_api.py (hunk 19 above)
Comment (dan1dr) [reply tests/module_1/test_meteo_api.py:10]: Tests implemented using MockResponses provided in the example code (please don't review this part). I need to fully digest the understanding and uses of mocking functions for future implementations. I modified the call_api functions and now all tests passed

**Review by dan1dr — COMMENTED** [review]
Todos los cambios solucionados. No dudes en comentarme cualquier nitpick que se te ocurre, para mi me interesa ver 100% cómo lo implementarías o mejor manera de abordar cualquier tonteria. He borrado la función de clean_data ya que no se usaba y he usado la mockresponse para call_api (cogiéndolo del código ejemplo), para ello he tenido q ajustar un par de cambios en call_api. Me faltaría digerir como puede implementartse Mocking en otro tipo de casos, ya que de primeras me cuesta verlo.

Gracias!!

File: src/module_1/module_1_meteo_api.py (hunk 8 above)
Comment (gsantam) [reply src/module_1/module_1_meteo_api.py]: Yes, exactly, DEBUG is the max verbosity. Normally in production one uses `INFO` but this parameter is normally parametrised in a `yaml` file so can be changed easily between executions / environments to a desired level. For example, if one wants to solve a problem, can change it to `DEBUG`

File: src/module_1/module_1_meteo_api.py [hunk 20]
```diff
@@ -2,167 +2,204 @@
...
+            else:
+                logging.error(f"\nAPI request failed with Code: {response.status_code}")
+        except requests.exceptions.RequestException as exception:
+            logging.error(
+                f"\nAPI request to url: {url} failed with Exception: {exception}"
+            )
+
+        retry_count += 1
+        if retry_count < MAX_RETRY_ATTEMPTS:
+            print(f"Retrying after {COOL_OFF_TIME} seconds...\n")
```
Comment (gsantam) [inline src/module_1/module_1_meteo_api.py:80]: Still some print here, use `logging.info`

File: src/module_1/module_1_meteo_api.py (hunk 11 above)
Comment (gsantam) [reply src/module_1/module_1_meteo_api.py]: Ideally, parametrize the INFO in order the log the name of the city/url too so you get extra info at a very low cost, so: `logging.info(f"API request connected successfully to {url}")` or `logging.error(f"\nAPI request failed with Code: {response.status_code}" for url {ur})`. Otherwise is gonna be hard to know which `url` is failing

**Review by gsantam — APPROVED** [review]
Tiene buena pinta después de los cambios, te he añadido un par de comentarios más, échales un ojo. Una cosa que veo que haces bastante es `return None`, esto no está mal pero es redundante, haz simplemente `return` (ya que return a secas SIEMPRE retorna None). Sobre lo de los mock functions si quieres lo comentamos por slack / llamada
